ui/src/text_annotator.tsx

```diff
@@ -315,17 +319,18 @@ export
           ]}
           />
         </div>
-        <div className={clas(css.content, 'wave-s16 wave-t7 wave-w3')}>{
+        <div className={clas(css.content, 'wave-s16 wave-t7 wave-w3')} onMouseUp={handleSelectionFinish}>{
           tokens.map(({ start, end, text, tag }, idx) => {
             const activeColor = activeTag ? tagColorMap.get(activeTag) : undefined
             return (
               // Use paragraph instead of span to support text highlight.
               <p
                 key={idx}
+                // data-keys keeps state in DOM to access it via window.getSelection() API in case mouse up event is fired when cursor is not above <p/> element
+                data-keys={`{ "key": ${idx}, "start": ${start}, "end": ${end} }`}
```

Comment:
Wouldn't this work?

```suggestion
                data-keys={{ idx, start, end }}
```

Reply:
It does not since dataset can be only string - https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset
The solution here is: 
- to use string representation of JSON and later parse it
- use separate `data-*` attributes for `idx`, `start` and `,end`